packages/block-editor/src/store/private-selectors.js

```diff
@@ -300,10 +300,10 @@ export const hasAllowedPatterns = createRegistrySelector( ( select ) =>
 				if ( ! inserter ) {
 					return false;
 				}
-				const { blocks } = getParsedPattern( pattern );
+				const grammar = getGrammar( pattern );
 				return (
-					checkAllowListRecursive( blocks, allowedBlockTypes ) &&
-					blocks.every( ( { name: blockName } ) =>
+					checkAllowListRecursive( grammar, allowedBlockTypes ) &&
```

Comment:
It's funny that `checkAllowListRecursive` is already made to work with the grammar objects... makes me wonder if we did that at some point before.

Reply:
It looks like it was used when the method was introduced - #30366. We should probably check why it was removed in case there was an issue with grammar parser.

Reply:
My guess is that there was a need elsewhere for parsed content, and someone switched it to a full parse. But yes we should try to fish it out of the history.

Reply:
It was removed in #32376.

Reply:
And it was removed to improve correctness of the block names: fallback names, deprecations that migrate to another block type etc.

Reply:
I'd like to understand better why it was changed, I don't understand it based on the issue and PR, so I left a [comment](https://github.com/WordPress/gutenberg/pull/32376#issuecomment-2317680334).

Reply:
@jsnajdr No, as far as I understand, it was removed for whitespace? That's what I see in the issue and also the testing instructions.

Reply:
I see, the issue is about whitespace, but the PR description says it's fixing a lot of other things, too: https://github.com/WordPress/gutenberg/pull/32376#issue-908391631

It's like if the lighweight parser had trouble with the whitespace. But that looks like an easily fixable bug.

Reply:
We even have block migrations in core that are done with a useEffect... It's impossible to catch all cases I think.

Reply:
Considering some of the edge cases we might miss. This might not be the best candidate for the minor WP release.

Reply:
I've been playing with `block-serialization-default-parser` a bit. When I pass a string with whitespace and without, the difference is that the whitespace triggers creation of empty blocks:
```
{
    blockName: null,
    attrs: {},
    innerBlocks: [],
    innerHTML: '\n\t',
    innerContent: [ '\n\t' ]
}
```
The wrapper in `@wordpress/blocks` then filters them out.

This happens only at top-level though:
```
<!-- wp:image --><figure><!-- /wp:image>[ whitespace ]<!-- wp:heading --><h2><!-- /wp:heading -->
```
the whitespace creates the `null`-ish block and the return value from the parser is array with 3 items.

If I wrap this in `wp:group`, then there are only two parsed inner blocks, the whitespace is ignored.

Reply:
So we can just filter out the empty things?

Reply:
Filtered them out. I'll see if I can add an e2e test.

Reply:
Added an e2e test to make sure #32376 does not regress.

Reply:
> So we can just filter out the empty things?

That's a good hotfix, but we should figure out why the underlying parser produces them in the first place and fix that. I don't see how they could possibly be useful. Especially when the same markup inside a nested block will not produce the empty blocks.

Reply:
> why the underlying parser produces them in the first place

I'm now looking at the code and it's totally intentional!

https://github.com/WordPress/gutenberg/blob/trunk/packages/block-serialization-default-parser/src/index.js#L273-L290

There is a token of type `void-block` and it produces block records only on top level. @dmsnell why does the parser do that? It looks like it has something to do with freeform/classic content.

Reply:
yes the goal of the parser is to represent the source document, so creating a fallback block for whitespace captures that and lets us write it back.

there was considerable discussion about this a number of years ago. the suggestion in each case is just to ignore the whitespace-only fallback blocks.

Reply:
@dmsnell do you remember why the fallback blocks with whitespace are created only on top level, but not in inner blocks?

If the goal of the parser is to represent the source document well, it should also capture whitespace in a nested markup.

Two nice things about using the lightweight parser for parsing patterns:
- patterns are more likely to have correct and modern block markup, because the concept of patterns is new. It's mostly the post content where dealing with legacy old things is important.
- if we ever implement block lazy loading, that means the full two-stage parse must be async (promise-returning) because block definitions are loaded by the parser as it needs them. But the first-state lighter parser can remain completely sync.

Reply:
> do you remember why the fallback blocks with whitespace are created only on top level, but not in inner blocks?

@jsnajdr are you certain what you are seeing is as you describe it? these aren't "empty blocks" so much as they are HTML content _outside_ of a block, or "HTML soup." the whitespace isn't ignored just because it's inside of another block, but when it's HTML content _inside_ of a block it becomes part of that block's inner HTML and inner content. in fact, this is _always_ the case because the whitespace is just `innerHTML`

You can explore this at my parser explorer - https://dmsnell.github.io/peg-parser-explorer/ - just make sure to click the <kbd>Fetch</kbd> button to load the block grammar.

`innerContent` is an array containing string chunks of HTML content, and `null` as a placeholder for inner blocks, allowing the reconstruction of multiple non-nested inner blocks within another block. all whitespace should appear both in `innerHTML` and in `innerContent`.

Reply:
Thanks for clarification @dmsnell, now I fully understand what's going on. At the top level the whitespace is represented as void blocks, because there is no `innerContent` field where we could put that information.

For inner blocks, the whitespace is in `innerContent`, which is an array that looks like `["\n\t", null, "\n\t", null, "\n"]`. The whitespace can be fully reconstructed from this. We don't need void blocks here.

There is also `innerHTML` field which is a string like `"\n\t\n\t\n"`. This field is only useful for "leaf" blocks that have their own markup that's in `innerHTML`, but for "container" blocks the information is not very useful, as all the whitespace is concatenated together.